// gzipper/src/compressions/Compression.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Compression = void 0;
var Compression = /** @class */ (function () {
    /**
     * Creates an instance of Compression.
     */
    function Compression(options, logger) {
        this.compressionOptions = {};
        this.options = options;
        this.logger = logger;
        this.selectCompression();
    }
    /**
     * Returns human-readable compression options info.
     */
    Compression.prototype.readableOptions = function (keyWrapper) {
        var e_1, _a;
        if (keyWrapper === void 0) { keyWrapper = function (key) { return key; }; }
        var options = '';
        try {
            for (var _b = __values(Object.entries(this.compressionOptions)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), key = _d[0], value = _d[1];
                options += keyWrapper(key) + ": " + value + ", ";
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return this.compressionName + " | " + options.slice(0, -2);
    };
    /**
     * Build compression options object [compressionOptions].
     */
    Compression.prototype.selectCompression = function () {
        var options = {};
        if (this.options.level !== undefined) {
            options.level = this.options.level;
        }
        if (this.options.memoryLevel !== undefined) {
            options.memLevel = this.options.memoryLevel;
        }
        if (this.options.strategy !== undefined) {
            options.strategy = this.options.strategy;
        }
        this.compressionOptions = options;
    };
    return Compression;
}());
exports.Compression = Compression;
